File: strategies/npm.go
// Copyright 2025 Naren Yellavula
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package strategies

// NpmHelpStrategy handles npm commands
type NpmHelpStrategy struct {
	cmdRunner *CommandRunner
}

func NewNpmHelpStrategy(cmdRunner *CommandRunner) *NpmHelpStrategy {
	return &NpmHelpStrategy{cmdRunner: cmdRunner}
}

func (n *NpmHelpStrategy) SupportsCommand(baseCmd string) bool {
	return baseCmd == "npm"
}

func (n *NpmHelpStrategy) Priority() int {
	return 2
}

func (n *NpmHelpStrategy) GetHelp(cmdParts []string) (string, error) {
	cmd := NewCommand(cmdParts)

	if !cmd.HasSubCommand(1) {
		return n.cmdRunner.Run("npm", "help")
	}

	subCmd := cmd.GetSubCommand(0)
	if out, err := n.cmdRunner.Run("npm", "help", subCmd); err == nil {
		return RemoveOverstrike(out), nil
	}

	// Fallback to npm <subcommand> --help
	return n.cmdRunner.Run("npm", subCmd, "--help")
}

File: recaller-rs/src/help_system/strategies/npm.rs
use std::sync::Arc;

use anyhow::Result;

use crate::help_system::runner::CommandRunner;
use crate::help_system::strategies::remove_overstrike;
use crate::help_system::strategy::{CommandParts, HelpStrategy};

pub struct NpmHelpStrategy {
    runner: Arc<CommandRunner>,
}

impl NpmHelpStrategy {
    pub fn new(runner: Arc<CommandRunner>) -> Self {
        Self { runner }
    }
}

impl HelpStrategy for NpmHelpStrategy {
    fn priority(&self) -> i32 {
        2
    }

    fn supports_command(&self, base_cmd: &str) -> bool {
        base_cmd == "npm"
    }

    fn get_help(&self, command: &CommandParts) -> Result<String> {
        if !command.has_sub_command(1) {
            return self.runner.run("npm", &["help"]);
        }

        let sub_cmd = command.get_sub_command(0).unwrap();
        if let Ok(out) = self.runner.run("npm", &["help", sub_cmd]) {
            return Ok(remove_overstrike(&out));
        }

        self.runner.run("npm", &[sub_cmd, "--help"])
    }
}
